runtime/reporter.rs

```diff
@@ -0,0 +1,60 @@
+use std::cell::RefCell;
+
+// Report events, activities and messages from the running module
+pub trait Reporter {
+    /// Print a debug log message. This is typically triggered by Console APIs like `console.log`.
+    /// Important: these messages include the final newline, the reporter SHOULD NOT add EOL.
+    fn debug_print(&self, msg: &str);
+
+    /// Record an activity log entry with level "info".
+    /// Important: this message DOES NOT include the final newline. The reporter should add EOL.
+    fn info_activity(&self, msg: &str);
+
+    /// Record an activity log entry with level "error".
+    /// Important: this message DOES NOT include the final newline. The reporter should add EOL.
+    fn error_activity(&self, msg: &str);
+
+    /// Report that module completed another job.
+    fn job_completed(&self);
+}
+
+/// Reporter that collects all recorded events, useful for testing.
+pub struct RecordingReporter {
+    pub events: RefCell<Vec<String>>,
```

Comment:
We want the recorder to be immutable so that we can pass it around and have multiple immutable references. However, our inner state is mutable. The solution is called "interior mutability", see https://doc.rust-lang.org/book/ch15-05-interior-mutability.html